### PR TITLE
VPA artificial samples generation for OOMs.

### PR DESCRIPTION
**Pull request by schylek**

### CODE REVIEW

File: vertical-pod-autoscaler/recommender/model/container.go [hunk 1]
```diff
@@ -126,6 +132,30 @@ func (container *ContainerState) addMemorySample(sample *ContainerUsageSample) b
...
+	// Discard old OOM
+	if timestamp.Before(container.WindowEnd.Add(-1 * MemoryAggregationInterval)) {
+		return fmt.Errorf("OOM event will be discarded - it is too old (%v)", timestamp)
+	}
+	// If OOM is in current or next window max requested mem with last known sample.
+	if timestamp.Before(container.WindowEnd.Add(MemoryAggregationInterval)) &&
+		container.MemoryUsagePeaks.Head() != nil {
+		resourceAmount = math.Max(resourceAmount, *container.MemoryUsagePeaks.Head())
+	}
+	resourceAmount *= OOMBumpUpRatio
```
Comment (mwielgus) [inline vertical-pod-autoscaler/recommender/model/container.go]: make it max(resourceAmount + 100MB, resourceAmount * OOMBumpRatio)

Comment (schylek) [reply vertical-pod-autoscaler/recommender/model/container.go]: Done.

File: vertical-pod-autoscaler/recommender/model/container.go [hunk 2]
```diff
@@ -126,6 +132,33 @@ func (container *ContainerState) addMemorySample(sample *ContainerUsageSample) b
...
+	if timestamp.Before(container.WindowEnd.Add(-1 * MemoryAggregationInterval)) {
+		return fmt.Errorf("OOM event will be discarded - it is too old (%v)", timestamp)
+	}
+	// If OOM is in current or next window max requested mem with last known sample.
+	if timestamp.Before(container.WindowEnd.Add(MemoryAggregationInterval)) &&
+		container.MemoryUsagePeaks.Head() != nil {
+		resourceAmount = math.Max(resourceAmount, *container.MemoryUsagePeaks.Head())
+	}
+
+	// Incresa by 100MB or 20%
```
Comment (bskiba) [inline vertical-pod-autoscaler/recommender/model/container.go]: s/Incresa/Increase

File: vertical-pod-autoscaler/recommender/model/container.go [hunk 3]
```diff
@@ -126,6 +132,33 @@ func (container *ContainerState) addMemorySample(sample *ContainerUsageSample) b
...
 }
 
+// RecordOOM adds info regarding OOM event in the model as an artifical memory sample.
+func (container *ContainerState) RecordOOM(timestamp time.Time, requestedMemory ResourceAmount) error {
+	resourceAmount := float64(requestedMemory)
+	// Discard old OOM
+	if timestamp.Before(container.WindowEnd.Add(-1 * MemoryAggregationInterval)) {
+		return fmt.Errorf("OOM event will be discarded - it is too old (%v)", timestamp)
+	}
+	// If OOM is in current or next window max requested mem with last known sample.
```
Comment (bskiba) [inline vertical-pod-autoscaler/recommender/model/container.go]: nit: I had to read the code anyway to understand this comment :)

File: vertical-pod-autoscaler/recommender/model/container.go [hunk 4]
```diff
@@ -126,6 +132,33 @@ func (container *ContainerState) addMemorySample(sample *ContainerUsageSample) b
...
+	if timestamp.Before(container.WindowEnd.Add(MemoryAggregationInterval)) &&
+		container.MemoryUsagePeaks.Head() != nil {
+		resourceAmount = math.Max(resourceAmount, *container.MemoryUsagePeaks.Head())
+	}
+
+	// Incresa by 100MB or 20%
+	resourceAmount = math.Max(resourceAmount+100*1024*1024, resourceAmount*OOMBumpUpRatio)
```
Comment (bskiba) [inline vertical-pod-autoscaler/recommender/model/container.go]: Can we have the 100 as a const somewhere?

File: vertical-pod-autoscaler/recommender/model/container.go (hunk 3 above)
Comment (schylek) [reply vertical-pod-autoscaler/recommender/model/container.go]: Done.

File: vertical-pod-autoscaler/recommender/model/container.go (hunk 4 above)
Comment (schylek) [reply vertical-pod-autoscaler/recommender/model/container.go]: Done.

**Pull request merged**